src/rust/bitbox02-rust-c/src/lib.rs

```diff
@@ -36,14 +36,70 @@ mod sha2;
 mod workflow;
 
 // Whenever execution reaches somewhere it isn't supposed to rust code will "panic". Our panic
-// handler will print the available information on the screen. If we compile with `panic=abort`
-// this code will never get executed.
+// handler will print the available information on the screen and over RTT. If we compile with
+// `panic=abort` this code will never get executed.
 #[cfg(not(test))]
 #[cfg(not(feature = "testing"))]
 #[cfg_attr(feature = "bootloader", allow(unused_variables))]
 #[panic_handler]
 fn panic(info: &core::panic::PanicInfo) -> ! {
+    ::util::log::log!("{}", info);
     #[cfg(feature = "firmware")]
     bitbox02_rust::print_debug!(0, "Error: {}", info);
     loop {}
 }
+
+#[no_mangle]
+pub extern "C" fn rust_rtt_init() {
+    ::util::log::rtt_init();
+}
+
+// SAFETY: The pointer must point to a null terminated string
+#[no_mangle]
+#[cfg_attr(not(feature = "rtt"), allow(unused))]
+pub extern "C" fn rust_log(ptr: *const ::util::c_types::c_char) {
+    #[cfg(feature = "rtt")]
+    {
+        if ptr.is_null() {
+            panic!("`ptr` must be a valid pointer");
+        }
+        let s = unsafe { core::ffi::CStr::from_ptr(ptr as _) };
+        let s = unsafe { core::str::from_utf8_unchecked(s.to_bytes()) };
+        ::util::log::rtt_target::rprintln!("{}", s);
+    }
+}
+
+/// This function is for debug purposes only!
+///
+/// Don't use this in production, the returned pointer is only valid until this function is called
+/// again. This function is not thread safe.
+///
+/// Returns a null terminated string, suitable for printing with printf in C.
+///
+/// Max `len` is 64. This function panics if `len` is to large.
+///
+/// Usage:
+///    uint8_t arr[2] = {1,2};
+///    util_log("%s", rust_dbg_hex(arr, sizeof(arr)));
+///
+#[no_mangle]
+#[cfg(feature = "rtt")]
+pub extern "C" fn rust_dbg_hex(
```

Comment:
Out of curiosity, why did you do this in Rust? Seems a bit over-complicated, the C-equivalent would probably look much simpler right?

Reply:
Because we have the `hex` crate in rust, and I lean towards rust. So if I can pick I choose rust over C.

I guess if I reuse `util_uint8_to_hex` it could be pretty short in C yes.

Reply:
util_uint8_to_hex is a wrapper around rust hex :joy: so you would in fact be using rust, just the wrapper would be simpler.